### ISSUE TITLE
Drop Trait blocked for variables given to the worker thread

### ISSUE THREAD

**Issue by Skasselbard** [BurntSushi/chan-signal#21]
I have a variable that I want to be properly destroyed after a SIGINT. Therefor I have implemented the Drop trait for the type of the variable, initialize it before starting the runner thread and give the thread a reference.
Unfortunately the drop() function will never be called.
Here is an example to illustrate the problem:
```rust
#[macro_use]
extern crate chan;
extern crate chan_signal;

use std::thread;
use chan_signal::Signal;

struct Thing{
    a_thing: u8
}

impl Drop for Thing {
    fn drop(&mut self){
        println!("Thing #{:?} dropped", self.a_thing);
    }
}
fn main() {
    let signal = chan_signal::notify(&[Signal::INT, Signal::TERM]);
    let (sdone, rdone) = chan::sync(0);

    {
        let thing1 = Thing{a_thing: 1};
        let thing2 = Thing{a_thing: 2};
        thread::spawn(move || run( thing2, sdone));
        chan_select! {
            signal.recv() -> signal => {
                println!("received signal: {:?}", signal)
            },
            rdone.recv() => {
                println!("Program completed normally.");
            }
        }
    }
}

fn run(thing:  Thing, _sdone: chan::Sender<()>) {
    let thing3 = Thing{a_thing: 3};
    loop {
    }
}
```
running this example will output:
```
received signal: Some(INT)
Thing #1 dropped
```
Expected would be:
```
received signal: Some(INT)
Thing #1 dropped
Thing #2 dropped
```
I would expect thing3.drop() not to be called, since its thread is terminated violently.

My rustc is at 1.19.0 and I tested the code on Kubuntu 16.04

I assume rust can't call the destructor safely because the inner state of the object might be inconsistent (e.g. maybe the runner thread already started to call the destructor and didn't finished entirely). So I don't assume that this issue will be fixed anytime soon, but I thought it might be of interest.

**Comment by BurntSushi** [comment]
What problem are you trying to solve by implementing a `Drop` constructor?

More generally, the very specific piece of code you've written here seems to be acting as I would expect. You're moving `thing2` into `run`, which never terminates, and therefore `thing2` is never dropped. I don't think this issue has anything to do with `chan-signal` specifically, but it's hard to know without knowing what problem you're trying to solve.

**Comment by Skasselbard** [comment]
I want to read and write the gpios of the raspberry pi. For that I have to write into files like it is dokumented [here](http://raspberrypiguide.de/howtos/raspberry-pi-gpio-how-to/), to export and unexport them.
I provided a ``GPIO`` struct in a separate [library](https://github.com/Skasselbard/rustpiIO) (a project for me to learn more about the raspi, as well as rust)
The struct implements the ``Drop`` trait where the gpio will be unexported.

For test purposes, I wanted to write a program which creates (export) a ``GPIO``, read out its value in an endless loop (which I placed in the worker thread), which would be terminated on a SIGINT. 
After termination I wanted the the ``GPIO`` struct to run regularly out of scope, so that it will be unexported in the destructor.

Your library seemed to fit my purposes and was easily included in my project. Your example seemed to be strait forward, but reading your answer, I might have miss interpreted the threading part (I haven't read about thread handling in rust yet). I assumed the worker thread to be terminated after the SIGINT was received, but in your answer you stated that it actually never terminates.

So in my particular case the thing structs are actually ``GPIO`` structs. The rest would be pretty much the same.  
I hope with that my intend got clearer. If not feel free to ask. 

Do you have any suggestion how to address this problem properly (if possible)?

**Comment by BurntSushi** [comment]
Take a look at your function:

```rust
fn run(thing:  Thing, _sdone: chan::Sender<()>) {
    let thing3 = Thing{a_thing: 3};
    loop {
    }
}
```

The destructor for `thing` is run at the end of `run`, but the end of `run` is impossible to reach because of `loop {}`, which is an infinite loop. That's all there is to it.

You need to find a way to terminate `run` gracefully, probably with another channel.

**Comment by Skasselbard** [comment]
I expected that the destructor of thing3 would never be hit. Thing2 was the one I expected to be destructed after receiving the signal.

But I can achieve my goal using your signals, entirely without threading, when I check the signals inside the loop like this (inspired by one of your examples):
```rust
fn main() {
    let signal = chan_signal::notify(&[Signal::INT, Signal::TERM]);
    let mut thing = Thing{a_thing: 1};
    loop {
        chan_select! {
            default => {},
            signal.recv() -> signal => {
                println!("received signal: {:?}", signal);
                break;
            },
        }
       //do something with the thing
    }
}
```

Thank you for your time and the work on signal handling with rust!

**Comment by BurntSushi** [comment]
@Skasselbard But when you call `run`, you're **moving** `thing2` into `run`, which means the `thing2` won't get dropped until the end of `run`. The relationship with `thing3` is that `thing2` will get dropped after `thing3`, but the scope of both extends to the end of `run`. If `run` never ends, then the destructors for `thing2` and `thing3` won't ever run.

Your stated solution seems fine to me. It corresponds to a [non-blocking receive](https://docs.rs/chan/0.1.19/chan/#example-non-blocking-sendsreceives).

**Comment by Skasselbard** [comment]
Yes there was my misunderstanding. I didn't realize that the run thread will never finish and that thing2's lifetime is altered by moving to the new thread.

Thanks again for your help.
